part/txn.go

```diff
@@ -157,101 +157,166 @@ func (txn *Txn[T]) cloneNode(n *header[T]) *header[T] {
 func (txn *Txn[T]) insert(root *header[T], key []byte, value T) (oldValue T, hadOld bool, newRoot *header[T]) {
 	fullKey := key
 
-	if root.size() == 0 && root.getLeaf() == nil {
-		txn.watches[root.watch] = struct{}{}
-		newRoot = newLeaf(txn.opts, key, key, value).self()
-		if newRoot.watch == nil {
-			newRoot.watch = make(chan struct{})
-		}
-		return
-	}
-
-	newRoot = txn.cloneNode(root)
-	this := newRoot
+	this := root
 	thisp := &newRoot
+
+	// Try to insert the key into the tree. If we find a free slot into which to insert
+	// it, we do it and return. If an existing node exists where the key should go, then
+	// we stop. 'this' points to that node, and 'thisp' to its memory location. It has
+	// not been cloned.
 	for {
-		if bytes.HasPrefix(key, this.prefix) {
-			key = key[len(this.prefix):]
-			if len(key) == 0 {
-				leaf := this.getLeaf()
-				if leaf != nil {
-					oldValue = leaf.value
-					hadOld = true
-				}
-				if this.isLeaf() {
-					// This is a leaf node and we just cloned it. Update the value.
-					leaf.value = value
-				} else {
-					// This is a non-leaf node, create/replace the existing leaf.
-					this.setLeaf(newLeaf(txn.opts, key, fullKey, value))
-				}
-				return
-			}
+		if this.isLeaf() {
+			// We've reached a leaf node, cannot go further.
+			break
+		}
 
-			child, idx := this.findIndex(key[0])
+		if !bytes.HasPrefix(key, this.prefix) {
+			break
+		}
 
-			if child == nil || child.prefix[0] != key[0] {
-				// No node for this key, add it.
-				if this.size()+1 > this.cap() {
-					// too small, need to promote to a larger size.
-					if !txn.opts.rootOnlyWatch || this == newRoot {
-						txn.watches[this.watch] = struct{}{}
-					}
-					this = this.promote(!txn.opts.rootOnlyWatch || this == newRoot)
-				} else {
-					// it fits, just need a clone.
-					this = txn.cloneNode(this)
-				}
-				this.insert(idx, newLeaf(txn.opts, key, fullKey, value).self())
-				*thisp = this
-				return
-			}
+		// Prefix matched. Consume it and go further.
+		key = key[len(this.prefix):]
+		if len(key) == 0 {
+			// Our key matches this node.
+			break
+		}
 
-			// Node exists, replace it with a clone and recurse into it.
-			child = txn.cloneNode(child)
-			thisp = &this.children()[idx]
-			*thisp = child
-			this = child
-		} else {
-			// Reached a node with a different prefix, split it.
-			newPrefix := commonPrefix(key, this.prefix)
+		// Clone the parent.
+		this = txn.cloneNode(this)
+		*thisp = this
 
-			this.prefix = this.prefix[len(newPrefix):]
-			key = key[len(newPrefix):]
+		child, idx := this.findIndex(key[0])
+		if child == nil {
+			// We've found a free slot where to insert the key.
+			if this.size()+1 > this.cap() {
+				// Node too small, promote it to the next size.
+				this = this.promote(!txn.opts.rootOnlyWatch || this == newRoot)
```

Comment:
I checked whether the `promote`d node inherits the watch channel of where it is promoted from, but `promote` seems to allocate a new one. Why?